blockchain/stake/staketx.go

```diff
@@ -1221,3 +1222,73 @@ func FindSpentTicketsInBlock(block *wire.MsgBlock) *SpentTicketsInBlock {
 		RevokedTickets: revocations,
 	}
 }
+
+// CreateRevocationFromTicket creates a revocation transaction for the provided
+// ticket.
+func CreateRevocationFromTicket(ticketHash *chainhash.Hash,
+	ticketMinOuts []*MinimalOutput, revocationTxFee dcrutil.Amount,
+	revocationTxVersion uint16, params *chaincfg.Params) (*wire.MsgTx, error) {
+
+	// Create the revocation transaction and add the input.  The only input for a
+	// revocation transaction is a ticket submission (OP_SSTX tagged) output.
+	revocationTx := wire.NewMsgTx()
+	const ticketSubmissionOutput = 0
+	const tree = wire.TxTreeStake
+	ticketSubmissionAmount := ticketMinOuts[ticketSubmissionOutput].Value
+	prevOut := wire.NewOutPoint(ticketHash, ticketSubmissionOutput, tree)
+	txIn := wire.NewTxIn(prevOut, ticketSubmissionAmount, []byte{})
+	revocationTx.AddTxIn(txIn)
+
+	// Get the ticket commitment output details.
+	isP2SH, payToHashes, amounts, _, _, _ := SStxStakeOutputInfo(ticketMinOuts)
+
+	// Calculate the revocation output values from this data.
+	const revocationSubsidy = 0
+	revocationOutputAmounts := CalculateRewards(amounts, ticketSubmissionAmount,
+		revocationSubsidy)
+
+	// Add all the SSRtx-tagged transaction outputs to the transaction after
+	// performing some validity checks.
+	feeApplied := false
+	for i, payToHash := range payToHashes {
+		// Ensure amount is in the valid range for monetary amounts.
+		if amounts[i] <= 0 || amounts[i] > dcrutil.MaxAmount {
+			str := fmt.Sprintf("invalid SSTx amount: 0 >= %v > %v", amounts[i] <= 0,
+				dcrutil.MaxAmount)
+			return nil, stakeRuleError(ErrSStxBadCommitAmount, str)
+		}
+
+		// Create a new script which pays to the provided address specified in
+		// the original ticket tx.
+		var ssrtxOutScript []byte
+		switch isP2SH[i] {
+		case false: // P2PKH
+			addr, err := stdaddr.NewAddressPubKeyHashEcdsaSecp256k1V0(payToHash,
+				params)
+			if err != nil {
+				return nil, err
+			}
+			_, ssrtxOutScript = addr.PayRevokeCommitmentScript()
+		case true: // P2SH
+			addr, err := stdaddr.NewAddressScriptHashV0FromHash(payToHash, params)
+			if err != nil {
+				return nil, err
+			}
+			_, ssrtxOutScript = addr.PayRevokeCommitmentScript()
+		}
+
+		// Add the txout to our SSRtx tx.
+		amt := revocationOutputAmounts[i]
+		if !feeApplied && int64(revocationTxFee) < amt {
```

Comment:
I considered this, and I'm not against adding it in, but I left it out for a couple of reasons:

- It's still enforced in consensus, and it's up to the caller to pass a correct fee (no change to as it is today with `createrawssrtx`)
- It's moving to a fee of zero shortly if the auto revocations agenda passes (and assuming it passes and is active, this will never get called to create version 1 revocations any longer)